### PR TITLE
Dashboard (Settings View) - Add, remove and view publisher logos 

### PR DESCRIPTION
**Pull request by BrittanyIRL**
## Summary
This PR adds functionality to the Dashboard settings view to add, remove and view publisher logos. 

## Relevant Technical Choices
- As laid out [in PR 3032](https://github.com/google/web-stories-wp/issues/3032#issuecomment-663494150), the settings API gives us an array of IDs for logos as well as the current default logo ID. We need to take those IDs and retrieve the image source from the media API. 
- This PR adds in the necessary media API w/ media context provider/consumer. 
  - When the initial fetch of settings info happens and `publisherLogoIds` are found, that will trigger `getMediaById` so that we can show publisher logos. 
  - When a new logo is added, first it goes to the media API where it gets added as a resource, when that is successful the returned id gets sent to the settings API to update the publisher logo ids. 
  - When a logo is removed, it's just removed from settings. I chose to mimic what happens when you remove a logo in the editor, which is to not delete the entire image from the media library. 
- I chose to deviate from the layout for the sake of consistency with the headers in the dashboard. Am now using the shared header for the dashboard. And I added in the shared `Layout` functionality to allow for scroll. This also adds in the side nav which had been overlooked on this view so far (whoops!) and hides it for smaller viewports. 

## To-do
- Some more robust error handling here would be beneficial. 
  - Need to define the best way to verify that the image sizing is appropriate per [brief] as well as file formatting. I'm blocking some via the upload, but the media API has its own restrictions. (https://docs.google.com/document/d/1Xs4uOZ3KCNlTMdjDT6q7cvTUpA1xk-kzrnVLQALltQk/edit#heading=h.g8j12riqi2ft)
  - Testing surrounding errorr-ing APIs since media and settings are codependent for publisher logos
- Updating default logo
  - How do we handle removing the current default publisher logo? I have currently blocked the ability to delete the default (no delete button if active) but that seems confusing. 
  - If a user has no logos yet and they upload two at once, do we set one as the default? 
- How do want to handle if a user has more than 100 publisher logos? Or really, how do we want to handle a few rows of logos? It feels extremely excessive in rendering? I've set the media api to it's max per page result (100) but if I have more than 100 publisher logo ids to go fetch it won't bring them all back. Do we want to set this up in a promise to grab multiple pages at once (we have something similar for tags and categories) but that just seems like a very chaotic user experience. 
- Design a loading view for file uploads 

## User-facing changes
- None, hidden behind a feature flag 

### In Storybook 
storybook publisher logo upload props
- Dashboard/Views/EditorSettings/PublisherLogo
  - 4 publisher logos render, 3 are removable. 
  - local uploads work too here but size isn't going to error. 
  - Added an `isLoading` prop for media that you can play with here. I'm sure the UI will change but it gets the point across. 
  - Added a `canUploadFile` prop that will hide the fileUpload when false. 
  - Added an `uploadError` prop that will render an error message concerning file uploads. 
- Dashboard/Components/FileUpload
  - Now has an `isLoading` knob to switch 

### In Dashboard, make sure `enableSettingsView` flag is true

mass upload
**Uploading**
1. User can add a publisher logo via the upload button. 
2. User can add a publisher logo via the drag and drop functionality. 
3. User can add multiple publisher logos at once via the upload button.
4. User can add multiple publisher logos at once via the drag and drop functionality. 
5. If a file you want to upload is more than 100 MB you get an error and the file does not upload. 
6. If you are not logged in as an admin, you do not get to see the upload logo option - just the existing logos. (you have to manually switch the `enableSettingsView` flag for this one to be visible). 

dashboard removing publisher logos
**Removing** 
1. User can click remove button on a publisher logo (just not the currently active logo which has no button) and see that the logo is removed from settings. 
2. User can keydown on 'enter' to remove a publisher logo (just not the currently active logo which has no button) and see that the logo is removed from settings. 

**Rendering**
1. Logos set in the editor will show up in the dashboard 

### Tests Created
- Karma tests for editorSettings updated to include render and remove functionality surrounding publisher logos (upload testing is tricky here? Got stuck on it, didn't feel worth the delay) 
- Unit tests updated for editorSettings, publisherLogos, media reducer,  setting reducer
---

Addresses #3483

### CODE REVIEW

File: assets/src/dashboard/app/api/useSettingsApi.js [hunk 1]
```diff
@@ -76,20 +76,39 @@ export default function useSettingsApi(
...
       try {
+        const query = {};
+        if (googleAnalyticsId) {
+          query.web_stories_ga_tracking_id = googleAnalyticsId;
+        }
+
+        if (publisherLogoIds) {
+          query.web_stories_publisher_logos = [
+            ...new Set([...state.publisherLogoIds, ...publisherLogoIds]),
+          ];
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/api/useSettingsApi.js:97]: When an update to the publisher logos on settings is requested, we have to signal the entire list of updates since the same key is used for removing logos. 

were we to update with just the new one(s) then the existing publisher logo ids would be replaced.

File: assets/src/dashboard/app/reducer/media.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  ADD_MEDIA_FAILURE: 'add_media_failure',
+  FETCH_MEDIA_SUCCESS: 'fetch_media_success',
+  FETCH_MEDIA_FAILURE: 'fetch_media_failure',
+};
+
+export const defaultMediaState = {
+  error: {},
+  isLoading: false,
+  uploadedMediaIds: [],
+  publisherLogos: {},
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/reducer/media.js]: I'm meh about this naming. Technically it's just media, but it's only getting used for publisherLogos and this seems easier to read to me.

File: assets/src/dashboard/app/views/editorSettings/googleAnalytics/index.js [hunk 3]
```diff
@@ -52,28 +52,30 @@ export const TEXT = {
...
+  const setAnalyticsId = useCallback((newId) => _setAnalyticsId(newId), []);
+
+  useEffect(() => {
+    setAnalyticsId(googleAnalyticsId);
+  }, [googleAnalyticsId, setAnalyticsId]);
+
   const handleCancelUpdateId = useCallback(() => {
     setAnalyticsId(googleAnalyticsId);
-  }, [googleAnalyticsId]);
+  }, [googleAnalyticsId, setAnalyticsId]);
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/editorSettings/googleAnalytics/index.js]: Realized that the googleAnalyticsId is coming through subsequently and never seeing the light of day through here and into the input. Setting up useEffect to make sure proper values are passed.

File: assets/src/dashboard/app/views/editorSettings/index.js [hunk 4]
```diff
@@ -21,53 +21,142 @@ import { __ } from '@wordpress/i18n';
...
-    ({ newGoogleAnalyticsId }) => {
-      const updatedSettings = {
-        googleAnalyticsId:
-          typeof newGoogleAnalyticsId === 'string' && newGoogleAnalyticsId,
-      };
+  useEffect(() => {
+    if (uploadedMediaIds.length > 0) {
+      updateSettings({ publisherLogoIds: uploadedMediaIds });
+    }
+  }, [updateSettings, uploadedMediaIds]);
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/editorSettings/index.js]: `uploadedMediaIds` is only set when new files are added as publisher logos. It's cleared when the upload starts and the Ids comeback on success. This shows us the proper id values to store as publisher logos in the settings API

File: assets/src/dashboard/app/views/editorSettings/index.js [hunk 5]
```diff
@@ -21,53 +21,142 @@ import { __ } from '@wordpress/i18n';
...
+    (newGoogleAnalyticsId) =>
+      updateSettings({ googleAnalyticsId: newGoogleAnalyticsId }),
+    [updateSettings]
+  );
+
+  const handleAddLogos = useCallback(
+    (files) => {
+      const isFileSizeWithinMaxUpload = files.every(
+        (file) => file.size <= maxUpload
+      );
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/editorSettings/index.js:93]: we could check locally here if the file size/dimensions are accurate. I'm not sure how much of this file checking we want to have happen in the front end.

File: includes/Dashboard.php [hunk 6]
```diff
@@ -261,7 +261,7 @@ public function get_dashboard_settings() {
 				'version'      => WEBSTORIES_VERSION,
 				'api'          => [
 					'stories'   => sprintf( '/web-stories/v1/%s', $rest_base ),
-					'media'     => '/web-stories/v1/media',
+					'media'     => '/wp/v2/media',
```
Comment (swissspidy) [inline includes/Dashboard.php]: Why was this changed? 🤔 

`/web-stories/v1/media` should be correct. It is the same as `/wp/v2/media` but also supports `?=_web_stories_envelope=1`

File: assets/src/dashboard/app/views/editorSettings/index.js [hunk 7]
```diff
@@ -21,53 +21,141 @@ import { __ } from '@wordpress/i18n';
...
+
+      if (!isFileSizeWithinMaxUpload) {
+        const errorText =
+          files.length > 1
```
Comment (swissspidy) [inline assets/src/dashboard/app/views/editorSettings/index.js]:
```suggestion
          files.length === 1
```

I think?

File: assets/src/dashboard/app/views/editorSettings/index.js [hunk 8]
```diff
@@ -21,53 +21,141 @@ import { __ } from '@wordpress/i18n';
...
+            ? __(
+                'Sorry, this file is too big. Make sure your logo is under 100 MB.',
+                'web-stories'
+              )
+            : __(
+                'Sorry, one or more of these files are too big. Make sure your logos are all under 100 MB.',
```
Comment (swissspidy) [inline assets/src/dashboard/app/views/editorSettings/index.js]: Where is this 100 MB limit coming from?

For these kind of values it's better to use `sprintf` with placeholders. So. `[...] under %s.`

File: assets/src/dashboard/app/api/useMediaApi.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        type: MEDIA_ACTION_TYPES.LOADING_MEDIA,
+      });
+
+      try {
+        const response = await dataAdapter.get(
+          queryString.stringifyUrl({
+            url: globalMediaApi,
+            query: {
+              include: mediaIds.join(','),
+              per_page: 100,
```
Comment (swissspidy) [inline assets/src/dashboard/app/api/useMediaApi.js:51]: We could theoretically do `per_page: -1` to load more than 100 items

Comment (swissspidy) [reply assets/src/dashboard/app/api/useMediaApi.js:51]: However, the database column for the publisher logos IDs in WordPress has some limit, so I don't think we'd ever reach such a high number.

File: includes/Dashboard.php (hunk 6 above)
Comment (BrittanyIRL) [reply includes/Dashboard.php]: ah, i changed it because of this: https://github.com/google/web-stories-wp/issues/3032#issuecomment-663494150 

I will swap it back!

File: assets/src/dashboard/app/api/useMediaApi.js (hunk 9 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/app/api/useMediaApi.js:51]: It seems like an odd case to have more than 100 publisher logos ever? We can loop through pages as we have set up for categories and tags. I'm also concerned about what this looks like on the page. Right now, how i have it set up if you have more than 100 it won't break, those just won't render.

Thoughts on leaving it as an unsolved edge case to put it on the back burner to get some UX about how to handle it so that it doesn't go unhandled in the long run?  I can make a bug ticket for it and tag as ux needed.

File: assets/src/dashboard/app/views/editorSettings/index.js (hunk 8 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/app/views/editorSettings/index.js]: ah thank you. I was using 100 MB because that's what's coming through as the maxUpload in config and what the other upload max message says natively in WordPress. That said, bit off too much in one PR and let this slide. Thanks for calling it out. Added in a util to convert the bytes received as maxUpload into a readable number - (`utils/formatBytes.js`) and updated translations to use sprintf. 

If you know a more graceful way to handle the plural/singular here I'm all ears.

File: includes/Dashboard.php (hunk 6 above)
Comment (BrittanyIRL) [reply includes/Dashboard.php]: Swapped and confirmed. 

Screen Shot 2020-08-18 at 8 29 25 AM
Screen Shot 2020-08-18 at 8 29 39 AM

File: assets/src/dashboard/app/api/useMediaApi.js (hunk 9 above)
Comment (swissspidy) [reply assets/src/dashboard/app/api/useMediaApi.js:51]: Yeah I don't think we need to accommodate for such a case.

File: assets/src/dashboard/utils/formatBytes.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// https://stackoverflow.com/questions/15900485/correct-way-to-convert-size-in-bytes-to-kb-mb-gb-in-javascript/18650828#18650828?newreg=85b17a31d1594ab7bdce436172642949
+const formatBytes = (bytes, decimals = 2) => {
```
Comment (swissspidy) [inline assets/src/dashboard/utils/formatBytes.js]: We should try to avoid copying & adapting code due to potential licensing issues.

Perhaps we could port [the PHP version of this utility from WordPress](https://developer.wordpress.org/reference/functions/size_format/#source) to JavaScript?

Feels more readable too as it doesn't use `log` or `pow`.

Interestingly, the units should be translatable too

Comment (BrittanyIRL) [reply assets/src/dashboard/utils/formatBytes.js]: Sure, didn't think about the units being converted. 

Added `maxUploadFormatted` to dashboard config using `use_size` instead - works as expected. Is this what you had in mind? 
`'maxUploadFormatted' => size_format( $max_upload_size ),`

Comment (swissspidy) [reply assets/src/dashboard/utils/formatBytes.js]: I was more talking about refactoring the JS util but this certainly easier 😄 👍 

Looks great!

File: assets/src/dashboard/app/reducer/media.js (hunk 2 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/app/reducer/media.js]: Figured out some better names :) Keeping purpose of media agnostic in the media reducer (which was bothering me). Update is incoming.

File: assets/src/dashboard/app/views/editorSettings/index.js [hunk 11]
```diff
@@ -15,60 +15,160 @@
...
   );
 
+  const orderedPublisherLogos = useMemo(() => {
+    if (Object.keys(mediaById).length <= 0) {
+      return [];
+    }
+    return publisherLogoIds.map((publisherLogoId) => {
+      if (mediaById[publisherLogoId]) {
+        if (publisherLogoId === activePublisherLogoId) {
+          mediaById[publisherLogoId].isActive = true;
```
Comment (dmmulroy) [inline assets/src/dashboard/app/views/editorSettings/index.js]: Mutating this will mutate the value back in the provider (and not trigger any updates) and could cause some confusing bugs.  Rather than mutating this here we should create an effect to update the provider state or we can do the following in the the return statement:

```javascript
return publisherLogoId === activePublisherLogoId ? { ...mediaById[publisherLogoId], isActive: true } : mediaById[publisherLogoId];
```

File: assets/src/dashboard/components/inlineInputForm/index.js [hunk 12]
```diff
@@ -47,7 +48,13 @@ const InlineInputForm = ({
   value,
 }) => {
   const inputContainerRef = useRef(null);
-  const [newValue, setNewValue] = useState(value);
+  const [newValue, _setNewValue] = useState(value);
+
+  const setNewValue = useCallback((newVal) => _setNewValue(newVal), []);
```
Comment (dmmulroy) [inline assets/src/dashboard/components/inlineInputForm/index.js]: This `useCallback` is unneeded since the setter reference will never change

**Review by dmmulroy — CHANGES_REQUESTED** [review]
Just some minor changes requested, but everything else looks good and works well 👍

File: assets/src/dashboard/app/views/editorSettings/index.js (hunk 11 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/app/views/editorSettings/index.js]: great catch, thanks! updating now 🙃

**Review by dmmulroy — APPROVED** [review]

**Review by mariano-formidable — APPROVED** [review]
Looks great!  I tested in dashboard, uploaded a bunch of logos and deleted them and refreshed and it all worked!

**Comment by carloskelly13** [comment]
@swissspidy For this do we want to have it QA'd or merge it in then create a separate ticket for QA. It's a huge feature rather than a couple of components behind a flag, so I don't want it to slip under the radar with our QA flow.

**Comment by swissspidy** [comment]
@carlos-kelly It's still behind a flag, right? So a new ticket to enable the flag & QA it is the way to go.

**Review by carloskelly13 — APPROVED** [review]
